kbkdf/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+///        as an integer in the interval [1, 2r − 1].
+pub trait Kbkdf<Prf, K, R: sealed::R = U32>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArrayLength<u8> + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArrayLength<u8> + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+{
+    /// Derives `key` from `kin` and other parameters.
+    fn derive(
+        &self,
+        kin: &GenericArray<u8, Prf::KeySize>,
+        use_l: bool,
+        use_separator: bool,
+        label: &[u8],
+        context: &[u8],
+    ) -> Result<GenericArray<u8, K::KeySize>, Error> {
+        // n - An integer whose value is the number of iterations of the PRF needed to generate L
+        // bits of keying material
+        let n: u32 = Wrapping(<KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::L::U32)
+            .div_ceil(Wrapping(
+                <KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::H::U32,
+            ))
+            .0;
+
+        if n as usize > 2usize.pow(R::U32) - 1 {
+            return Err(Error::InvalidRequestSize);
+        }
+
+        let mut output = GenericArray::<u8, K::KeySize>::default();
+        let mut builder = output.as_mut_slice();
+
+        let mut ki = None;
+        self.input_iv(&mut ki);
+
+        for counter in 1..=n {
+            let mut h = <Prf as Mac>::new(kin);
+
+            if Self::FEEDBACK_KI {
+                if let Some(ki) = ki {
+                    h.update(ki.as_slice());
+                }
+            }
+
+            // counter encoded as big endian u32
+            // Type parameter R encodes how large the value is to be (either U8, U16, U24, or U32)
+            //
+            // counter = 1u32 ([0, 0, 0, 1])
+            //                     \-------/
+            //                      R = u24
+            h.update(&counter.to_be_bytes()[(4 - R::USIZE / 8)..]);
+
+            // Fixed input data
+            h.update(label);
+            if use_separator {
+                h.update(&[0]);
+            }
+            h.update(context);
+            if use_l {
+                h.update(
+                    &(<KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::L::U32).to_be_bytes()
+                        [..],
+                );
+            }
+
+            let buf = h.finalize().into_bytes();
+            ki = Some(buf.clone());
+
+            let remaining = usize::min(buf.len(), builder.len());
+
+            builder[..remaining].copy_from_slice(&buf[..remaining]);
+            builder = &mut builder[remaining..];
+        }
+
+        assert_eq!(builder.len(), 0, "output has uninitialized bytes");
+
+        Ok(output)
+    }
+
+    /// Input the IV in the PRF
+    fn input_iv(&self, _ki: &mut Option<GenericArray<u8, Prf::OutputSize>>) {}
+
+    /// Whether the KI should be reinjected every round.
+    const FEEDBACK_KI: bool = false;
+}
+
+pub struct Counter<Prf, K, R = U32> {
```

Comment:
not so fan of the default values for `R`. I wonder if maybe a type alias might not be better. (ala [`x509-cert::Certificate`](https://docs.rs/x509-cert/latest/x509_cert/certificate/type.Certificate.html))